File: services/agent-mgmt/app/adapters/redisAdapter.ts
import Redis from 'ioredis';
import { v4 as uuidv4 } from 'uuid';
import { config } from '../config';
import { logger } from '../utils/logger';

const STREAM_MAXLEN = 10000;
const SOURCE_SERVICE = 'agent-mgmt';
const SCHEMA_VERSION = '1';

class RedisAdapter {
  private client: Redis | null = null;

  connect(): void {
    this.client = new Redis(config.REDIS_URL, {
      retryStrategy: (times) => Math.min(times * 100, 3000),
      maxRetriesPerRequest: 3,
      enableReadyCheck: true,
      lazyConnect: false,
    });

    this.client.on('connect', () => {
      logger.info('Redis connected', 'adapter', { url: config.REDIS_URL });
    });

    this.client.on('error', (err: Error) => {
      logger.error('Redis error', 'adapter', { error: err.message });
    });

    this.client.on('reconnecting', () => {
      logger.warn('Redis reconnecting', 'adapter');
    });
  }

  getClient(): Redis {
    if (!this.client) {
      throw new Error('Redis client not initialized — call connect() first');
    }
    return this.client;
  }

  async publish(stream: string, message: Record<string, unknown>): Promise<string> {
    const client = this.getClient();
    const enriched = {
      trace_id: (message.trace_id as string) || uuidv4(),
      source_service: SOURCE_SERVICE,
      timestamp: new Date().toISOString(),
      schema_version: SCHEMA_VERSION,
      ...message,
    };

    // XADD <stream> MAXLEN ~ <maxlen> * <fields...>
    const fields: string[] = [];
    for (const [key, value] of Object.entries(enriched)) {
      fields.push(key, typeof value === 'string' ? value : JSON.stringify(value));
    }

    const id = await client.xadd(stream, 'MAXLEN', '~', STREAM_MAXLEN, '*', ...fields);
    logger.debug(`Published to stream ${stream}`, 'adapter', {
      stream,
      message_id: id,
      trace_id: enriched.trace_id,
    });
    return id as string;
  }

  async healthCheck(): Promise<boolean> {
    try {
      const result = await this.getClient().ping();
      return result === 'PONG';
    } catch {
      return false;
    }
  }

  async close(): Promise<void> {
    if (this.client) {
      await this.client.quit();
      this.client = null;
      logger.info('Redis connection closed', 'adapter');
    }
  }
}

export const redisAdapter = new RedisAdapter();

File: services/orchestrator/app/adapters/redis_adapter.py
"""Redis async streams adapter for the orchestrator service."""

from __future__ import annotations

import json
import time
import uuid
from typing import Any, Callable, Dict, List, Optional

import redis.asyncio as aioredis

from app.config import settings
from app.utils.logger import logger

SCHEMA_VERSION = "1"


class RedisAdapter:
    """Wraps redis-py async client for stream operations."""

    def __init__(self) -> None:
        self._client: Optional[aioredis.Redis] = None

    async def connect(self) -> None:
        self._client = aioredis.from_url(
            settings.redis_url, encoding="utf-8", decode_responses=True
        )
        logger.info("Redis client connected", layer="adapter", url=settings.redis_url)

    async def close(self) -> None:
        if self._client:
            await self._client.aclose()
            logger.info("Redis client closed", layer="adapter")

    async def health_check(self) -> bool:
        try:
            await self._client.ping()
            return True
        except Exception as exc:
            logger.error("Redis health check failed", layer="adapter", error=str(exc))
            return False

    # ------------------------------------------------------------------
    # Stream producer
    # ------------------------------------------------------------------

    async def publish_to_stream(
        self,
        stream: str,
        message: Dict[str, Any],
        trace_id: str = "",
        maxlen: int = 10000,
    ) -> str:
        """XADD stream MAXLEN ~ maxlen * flat_message_dict. Returns message id."""
        enriched = {
            **message,
            "trace_id": trace_id or str(uuid.uuid4()),
            "source_service": settings.service_name,
            "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            "schema_version": SCHEMA_VERSION,
        }
        # Redis stream requires flat string values
        flat = {k: json.dumps(v) if not isinstance(v, str) else v for k, v in enriched.items()}
        msg_id = await self._client.xadd(stream, flat, maxlen=maxlen, approximate=True)
        logger.debug(
            "Published to Redis stream",
            layer="adapter",
            stream=stream,
            msg_id=msg_id,
            trace_id=trace_id,
        )
        return msg_id

    async def publish_memory_write(
        self,
        agent_id: int,
        tier: str,
        content: str,
        task_id: str,
        importance: float = 0.5,
        trace_id: str = "",
        task_description: str = "",
        score: Optional[float] = None,
        outcome: str = "SUCCESS",
    ) -> str:
        metadata: Dict[str, Any] = {
            "task_id": task_id,
            "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            "importance": importance,
            "outcome": outcome,
        }
        if task_description:
            metadata["task_description"] = task_description[:2000]
        if score is not None:
            metadata["score"] = score
        return await self.publish_to_stream(
            "memory:writes",
            {
                "agent_id": str(agent_id),
                "tier": tier,
                "content": content,
                "metadata": json.dumps(metadata),
            },
            trace_id=trace_id,
        )

    async def publish_telemetry(self, event: Dict[str, Any], trace_id: str = "") -> str:
        return await self.publish_to_stream("events:telemetry", event, trace_id=trace_id)

    async def publish_task(self, task: Dict[str, Any], trace_id: str = "") -> str:
        return await self.publish_to_stream("orchestrator:tasks", task, trace_id=trace_id)

    # ------------------------------------------------------------------
    # Stream consumer (scoring:feedback)
    # ------------------------------------------------------------------

    async def consume_scoring_feedback(
        self,
        group: str,
        consumer: str,
        handler: Callable[[Dict[str, Any]], None],
        batch_size: int = 10,
        block_ms: int = 5000,
    ) -> None:
        """
        Blocking XREADGROUP loop. Calls handler for each message.
        Caller is responsible for running this in a background task.
        """
        stream = "scoring:feedback"
        try:
            await self._client.xgroup_create(stream, group, id="$", mkstream=True)
        except Exception:
            pass  # group already exists

        while True:
            try:
                results = await self._client.xreadgroup(
                    group,
                    consumer,
                    {stream: ">"},
                    count=batch_size,
                    block=block_ms,
                )
                if not results:
                    continue
                for _, messages in results:
                    for msg_id, data in messages:
                        try:
                            parsed = {k: self._try_json(v) for k, v in data.items()}
                            await handler(parsed)
                            await self._client.xack(stream, group, msg_id)
                        except Exception as exc:
                            logger.error(
                                "Error processing scoring feedback message",
                                layer="adapter",
                                msg_id=msg_id,
                                error=str(exc),
                            )
            except Exception as exc:
                logger.error(
                    "Redis XREADGROUP error",
                    layer="adapter",
                    stream=stream,
                    error=str(exc),
                )
                # brief pause before retry
                import asyncio
                await asyncio.sleep(1)

    @staticmethod
    def _try_json(value: str) -> Any:
        try:
            return json.loads(value)
        except (json.JSONDecodeError, TypeError):
            return value
